query/src/main/scala/filodb/query/exec/HistogramQuantileMapper.scala

```diff
@@ -0,0 +1,154 @@
+import filodb.query.{Query, QueryConfig}
+
+object HistogramQuantileMapper {
+  import ZeroCopyUTF8String._
+  val le = "le".utf8
+}
+
+/**
+  * Calculates histogram quantile for one or more histograms whose bucket range vectors are passed
+  * into the apply function.
+  *
+  * @param funcParams Needs one double quantile argument
+  */
+case class HistogramQuantileMapper(funcParams: Seq[Any]) extends RangeVectorTransformer {
+
+  import HistogramQuantileMapper._
+  require(funcParams.size == 1, "histogram_quantile function needs a single quantile argument")
+
+  private val quantile = funcParams.head.asInstanceOf[Number].doubleValue()
+
+  /**
+    * Represents a prometheus histogram bucket for quantile calculation purposes.
+    * @param le the less-than-equals boundary for histogram bucket
+    * @param rate number of occurrences for the bucket per second
+    */
+  case class Bucket(val le: Double, var rate: Double) {
+    override def toString: String = s"$le->$rate"
+  }
+
+  /**
+    * Groups incoming bucket range vectors by histogram name. It then calculates quantile for each histogram
+    * using the buckets supplied for it. It is assumed that each bucket value contains rate of increase for that
+    * bucket.
+    *
+    * Important Note: The source range vectors for each bucket should NOT be the counter values themselves,
+    * but should be the rate of increase for that bucket counter. The histogram_quantile function should always
+    * be preceded by a rate function or a sum-of-rate function.
+    */
+  override def apply(source: Observable[RangeVector],
+                     queryConfig: QueryConfig, limit: Int,
+                     sourceSchema: ResultSchema): Observable[RangeVector] = {
+    val res = source.toListL.map { rvs =>
+
+      // first group the buckets by histogram
+      val histograms = groupRangeVectorsByHistogram(rvs)
+
+      // calculate quantile for each bucket
+      val quantileResults = histograms.map { histBuckets =>
+
+        // sort the bucket range vectors by increasing le tag value
+        val sortedBucketRvs = histBuckets._2.toArray.map { bucket =>
+          val leStr = bucket.key.labelValues(le).toString
+          val leDouble = if (leStr == "+Inf") Double.PositiveInfinity else leStr.toDouble
+          leDouble -> bucket
+        }.sortBy(_._1)
+
+        val samples = sortedBucketRvs.map(_._2.rows)
+
+        // The buckets here will be populated for each instant for quantile calculation
+        val buckets = sortedBucketRvs.map { b => Bucket(b._1, 0d) }
+
+        // create the result iterator that lazily produces quantile for each timestamp
+        val quantileResult = new Iterator[RowReader] {
+          val row = new TransientRow()
+          override def hasNext: Boolean = samples.forall(_.hasNext)
+          override def next(): RowReader = {
+            for { i <- 0 until samples.length } {
+              val nxt = samples(i).next()
```

Comment:
This will only work if the input RVs all have aligned timestamps.  Is this guaranteed?  Otherwise you are processing data points from possibly different timestamps.

Reply:
histogram quantile is always applied on periodic samples after rate is done. So timestamps will be aligned.